File: benchmarks/bench_oop.pl
#!/usr/bin/perl
use strict;
use warnings;

package Point;

sub new {
    my ($class, %args) = @_;
    return bless {
        x => $args{x} // 0,
        y => $args{y} // 0,
    }, $class;
}

sub x { return $_[0]->{x} }
sub y { return $_[0]->{y} }
sub set_x { $_[0]->{x} = $_[1] }
sub set_y { $_[0]->{y} = $_[1] }

sub distance_sq {
    my ($self) = @_;
    return $self->{x} * $self->{x} + $self->{y} * $self->{y};
}

package Point3D;
our @ISA = ('Point');

sub new {
    my ($class, %args) = @_;
    my $self = Point::new($class, %args);
    $self->{z} = $args{z} // 0;
    return $self;
}

sub z { return $_[0]->{z} }
sub set_z { $_[0]->{z} = $_[1] }

sub distance_sq {
    my ($self) = @_;
    return $self->{x} * $self->{x} + $self->{y} * $self->{y} + $self->{z} * $self->{z};
}

package main;

# Create 500,000 Point objects, call method on each
my $sum = 0;
my $i = 0;
while ($i < 500000) {
    my $p = Point->new(x => $i % 100, y => ($i + 1) % 100);
    $sum += $p->distance_sq();
    $i++;
}
print "point sum: $sum\n";

# Create 500,000 Point3D objects (inherited), call overridden method
my $sum3d = 0;
my $j = 0;
while ($j < 500000) {
    my $p = Point3D->new(x => $j % 100, y => ($j + 1) % 100, z => ($j + 2) % 100);
    $sum3d += $p->distance_sq();
    $j++;
}
print "point3d sum: $sum3d\n";

# isa() checks on 200,000 objects
my $isa_count = 0;
my $k = 0;
while ($k < 200000) {
    my $p = Point3D->new(x => 1, y => 2, z => 3);
    if ($p->isa("Point")) {
        $isa_count++;
    }
    $k++;
}
print "isa checks: $isa_count\n";

File: benchmarks/bench_oop.go
// Go counterpart of bench_oop.strada. Go has no class inheritance; the
// Point/Point3D relationship uses struct embedding with a method override,
// and the isa() check uses a runtime type assertion.
package main

import "fmt"

type Point struct{ x, y int }

func (p Point) distanceSq() int { return p.x*p.x + p.y*p.y }

type Point3D struct {
	Point
	z int
}

func (p Point3D) distanceSq() int { return p.x*p.x + p.y*p.y + p.z*p.z }

func main() {
	s := 0
	for i := 0; i < 500000; i++ {
		p := Point{i % 100, (i + 1) % 100}
		s += p.distanceSq()
	}
	fmt.Printf("point sum: %d\n", s)

	s3d := 0
	for j := 0; j < 500000; j++ {
		p := Point3D{Point{j % 100, (j + 1) % 100}, (j + 2) % 100}
		s3d += p.distanceSq()
	}
	fmt.Printf("point3d sum: %d\n", s3d)

	isaCount := 0
	for k := 0; k < 200000; k++ {
		p := Point3D{Point{1, 2}, 3}
		var iface interface{} = p
		if _, ok := iface.(Point3D); ok {
			isaCount++
		}
	}
	fmt.Printf("isa checks: %d\n", isaCount)
}
